cpp/src/text/vocabulary_tokenize.cu

```diff
@@ -0,0 +1,258 @@
+                                                           cudf::size_type,
+                                                           cuco::experimental::extent<std::size_t>,
+                                                           cuda::thread_scope_device,
+                                                           vocab_equal,
+                                                           probe_scheme,
+                                                           hash_table_allocator_type>;
+}  // namespace
+}  // namespace detail
+
+// since column_device_view::create returns is a little more than
+// std::unique_ptr<column_device_view> this helper simplifies the return type in a maintainable way
+using col_device_view = std::invoke_result_t<decltype(&cudf::column_device_view::create),
+                                             cudf::column_view,
+                                             rmm::cuda_stream_view>;
+
+struct tokenize_vocabulary::tokenize_vocabulary_impl {
+  std::unique_ptr<cudf::column> const vocabulary;
+  col_device_view const d_vocabulary;
+  std::unique_ptr<detail::vocabulary_map_type> vocabulary_map;
+
+  auto get_map_ref() const { return vocabulary_map->ref(cuco::experimental::op::find); }
+
+  tokenize_vocabulary_impl(std::unique_ptr<cudf::column>&& vocab,
+                           col_device_view&& d_vocab,
+                           std::unique_ptr<detail::vocabulary_map_type>&& map)
+    : vocabulary(std::move(vocab)), d_vocabulary(std::move(d_vocab)), vocabulary_map(std::move(map))
+  {
+  }
+};
+
+struct key_pair {
+  __device__ auto operator()(cudf::size_type idx) const noexcept
+  {
+    return cuco::make_pair(idx, idx);
+  }
+};
+
+tokenize_vocabulary::tokenize_vocabulary(cudf::strings_column_view const& input,
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(not input.is_empty(), "vocabulary must not be empty");
+  CUDF_EXPECTS(not input.has_nulls(), "vocabulary must not have nulls");
+
+  // need to hold a copy of the input
+  auto vocabulary   = std::make_unique<cudf::column>(input.parent(), stream, mr);
+  auto d_vocabulary = cudf::column_device_view::create(vocabulary->view(), stream);
+
+  auto vocab_map = std::make_unique<detail::vocabulary_map_type>(
+    static_cast<size_t>(vocabulary->size() * 2),
+    cuco::empty_key{-1},
+    cuco::empty_value{-1},
+    detail::vocab_equal{*d_vocabulary},
+    detail::probe_scheme{detail::vocab_hasher{*d_vocabulary}},
+    detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+    stream.value());
+
+  // the row index is the token id (value for each key in the map)
+  auto iter = cudf::detail::make_counting_transform_iterator(0, key_pair{});
+  vocab_map->insert_async(iter, iter + vocabulary->size(), stream.value());
+
+  _impl = new tokenize_vocabulary_impl(
```

Comment:
Why are these held by an "impl" rather than held as members of the `tokenize_vocabulary` class itself? Just because they take pre-processing work to initialize?

Reply:
Okay, that makes sense. I forgot that was in a public header.